light-prover/prover/batch_append_circuit.go

```diff
@@ -0,0 +1,364 @@
+//
+// The function works as follows:
+//  1. It starts with a carry of 1 (equivalent to adding 1 to the number).
+//  2. For each bit, from least to most significant:
+//     a. It XORs the current bit with the carry. This effectively adds the bit and carry
+//     without considering a new carry. (0⊕0=0, 0⊕1=1, 1⊕0=1, 1⊕1=0)
+//     b. It ANDs the original bit with the carry to determine if there should be a carry
+//     for the next bit. (0∧0=0, 0∧1=0, 1∧0=0, 1∧1=1)
+//     c. The result of XOR becomes the new value for the current bit.
+//     d. The result of AND becomes the new carry for the next iteration.
+//  3. This process continues for all bits, resulting in the incremented binary number.
+//
+// Example: Incrementing 0111 (7 in decimal)
+// Initial state: 0111, carry = 1
+// i=0: 1⊕1=0, carry=1∧1=1 -> 0110, carry=1
+// i=1: 1⊕1=0, carry=1∧1=1 -> 0010, carry=1
+// i=2: 1⊕1=0, carry=1∧1=1 -> 1010, carry=1
+// i=3: 0⊕1=1, carry=0∧1=0 -> 1000, carry=0
+// Final result: 1000 (8 in decimal)
+func (circuit *BatchAppendCircuit) incrementBits(api frontend.API, bits []frontend.Variable) []frontend.Variable {
+	carry := frontend.Variable(1)
+	for i := 0; i < len(bits); i++ {
+		// XOR operation implements binary addition without carry
+		newBit := api.Xor(bits[i], carry)
+		// AND operation determines if we need to carry to the next bit
+		carry = api.And(bits[i], carry)
+		bits[i] = newBit
+	}
+	return bits
+}
+
+func (circuit *BatchAppendCircuit) getZeroValue(api frontend.API, level int) frontend.Variable {
+	return frontend.Variable(new(big.Int).SetBytes(ZERO_BYTES[level][:]))
+}
+
+func (circuit *BatchAppendCircuit) createHashChain(api frontend.API, length int, hashes []frontend.Variable) frontend.Variable {
+	if length == 0 {
+		return 0
+	}
+
+	hashChain := hashes[0]
+	for i := 1; i < length; i++ {
+		hashChain = abstractor.Call(api, poseidon.Poseidon2{In1: hashChain, In2: hashes[i]})
+	}
+	return hashChain
+}
+
+type BatchAppendParameters struct {
+	OldSubTreeHashChain *big.Int   `json:"oldSubTreeHashChain"`
+	NewSubTreeHashChain *big.Int   `json:"newSubTreeHashChain"`
+	NewRoot             *big.Int   `json:"newRoot"`
+	HashchainHash       *big.Int   `json:"hashchainHash"`
+	StartIndex          uint32     `json:"startIndex"`
+	Leaves              []*big.Int `json:"leaves"`
+	Subtrees            []*big.Int `json:"subtrees"`
+	HashChainStartIndex uint32     `json:"hashChainStartIndex"`
+	TreeHeight          uint32     `json:"treeHeight"`
+	tree                *merkle_tree.PoseidonTree
+}
+
+func (p *BatchAppendParameters) TreeDepth() uint32 {
+	return p.TreeHeight
+}
+
+func (p *BatchAppendParameters) BatchSize() uint32 {
+	return uint32(len(p.Leaves))
+}
+
+func (p *BatchAppendParameters) ValidateShape(treeDepth uint32, batchSize uint32) error {
+	if p.TreeDepth() != treeDepth {
+		return fmt.Errorf("wrong tree depth: expected %d, got %d", treeDepth, p.TreeDepth())
+	}
```

Comment:
you are using `Depth` here while you renamed `Depth` to `Height` in many other places.